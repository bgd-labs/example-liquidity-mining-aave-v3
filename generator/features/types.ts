import {Hex} from 'viem';

export interface LiquidityMiningSetup {
  emissionsAdmin: Hex;
  rewardToken: string;
  rewardTokenDecimals: number;
  rewardOracle: string;
  assets: string[];
  distributionEnd: string;
  transferStrategy: Hex;
  rewardAmounts: string[];
  totalReward: number;
  whaleAddresses: Hex[];
  whaleExpectedRewards: string[];
}

export interface LiquidityMiningUpdate {
  emissionsAdmin: Hex;
  rewardToken: string;
  rewardTokenDecimals: number;
  rewardAmount: string;
  asset: string;
  distributionEnd: string;
  whaleAddress: Hex;
  whaleExpectedReward: string;
}
