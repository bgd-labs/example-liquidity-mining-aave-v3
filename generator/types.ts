import * as addressBook from '@bgd-labs/aave-address-book';
import {LiquidityMiningSetup, LiquidityMiningUpdate} from './features/types';

export const V3_POOLS = [
  'AaveV3Ethereum',
  'AaveV3EthereumLido',
  'AaveV3EthereumEtherFi',
  'AaveV3Polygon',
  'AaveV3Avalanche',
  'AaveV3Optimism',
  'AaveV3Arbitrum',
  'AaveV3Metis',
  'AaveV3Base',
  'AaveV3Gnosis',
  'AaveV3Scroll',
  'AaveV3BNB',
  'AaveV3ZkSync',
] as const satisfies readonly (keyof typeof addressBook)[];

export const POOLS = [...V3_POOLS] as const satisfies readonly (keyof typeof addressBook)[];

export type PoolIdentifier = (typeof POOLS)[number];
export type PoolIdentifierV3 = (typeof V3_POOLS)[number];

export interface Options {
  force?: boolean;
  feature: FEATURE;
  pool: PoolIdentifier;
  title: string;
  shortName: string;
  configFile?: string;
  date: string;
}

export type PoolConfigs = Partial<Record<PoolIdentifier, PoolConfig>>;

export type CodeArtifact = {
  code?: {
    constants?: string[];
    fn?: string[];
    execute?: string[];
  };
};

export enum FEATURE {
  SETUP_LM = 'SETUP_LM',
  UPDATE_LM = 'UPDATE_LM',
}

export interface FeatureModule<T extends {} = {}> {
  description: string;
  value: FEATURE;
  cli: (args: {options: Options; pool: PoolIdentifier; cache: PoolCache}) => Promise<T>;
  build: (args: {options: Options; pool: PoolIdentifier; cache: PoolCache; cfg: T}) => CodeArtifact;
}

export type ConfigFile = {
  rootOptions: Options;
  poolOptions: Partial<Record<PoolIdentifier, Omit<PoolConfig, 'artifacts'>>>;
};

export type PoolCache = {blockNumber: number};

export interface PoolConfig {
  artifacts: CodeArtifact[];
  configs: {
    [FEATURE.SETUP_LM]?: LiquidityMiningSetup;
    [FEATURE.UPDATE_LM]?: LiquidityMiningUpdate;
  };
  cache: PoolCache;
}

export type Files = {
  jsonConfig: string;
  payloadTest: {pool: PoolIdentifier; payloadTest: string; contractName: string};
};
