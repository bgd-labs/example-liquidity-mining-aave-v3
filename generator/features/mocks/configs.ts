import {Options, FEATURE} from '../../types';
import {LiquidityMiningSetup, LiquidityMiningUpdate} from '../types';

export const MOCK_OPTIONS_SETUP: Options = {
  pool: 'AaveV3EthereumLido',
  title: 'test',
  shortName: 'Test',
  date: '20231023',
  feature: FEATURE.SETUP_LM,
};

export const MOCK_OPTIONS_UPDATE: Options = {
  pool: 'AaveV3EthereumLido',
  title: 'test',
  shortName: 'Test',
  date: '20231023',
  feature: FEATURE.UPDATE_LM,
};

export const liquidityMiningSetupConfig: LiquidityMiningSetup = {
  emissionsAdmin: '0xac140648435d03f784879cd789130F22Ef588Fcd',
  rewardToken: 'AaveV3EthereumLidoAssets.wstETH_A_TOKEN',
  rewardTokenDecimals: 18,
  rewardOracle: 'AaveV3EthereumLidoAssets.wstETH_ORACLE',
  assets: ['wstETH_aToken'],
  distributionEnd: '14',
  transferStrategy: '0x0605a898535E9116Ff820347c536E3442F216Eb8',
  rewardAmounts: ['100'],
  totalReward: 100,
  whaleAddresses: ['0x07833EAdF87CD3079da281395f2fBA24b61F90f7'],
  whaleExpectedRewards: ['21.6'],
};

export const liquidityMiningUpdateConfig: LiquidityMiningUpdate = {
  emissionsAdmin: '0xac140648435d03f784879cd789130F22Ef588Fcd',
  rewardToken: 'AaveV3EthereumLidoAssets.wstETH_A_TOKEN',
  rewardTokenDecimals: 18,
  asset: 'wstETH_aToken',
  distributionEnd: '20',
  rewardAmount: '150',
  whaleAddress: '0x07833EAdF87CD3079da281395f2fBA24b61F90f7',
  whaleExpectedReward: '32.4',
};
