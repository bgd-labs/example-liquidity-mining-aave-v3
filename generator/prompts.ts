// PROMPTS
interface GenericPrompt<T extends boolean = boolean> {
  message: string;
  disableKeepCurrent?: T;
  transform?: (value: string) => string;
  defaultValue?: string;
}

interface PercentInputPrompt<T extends boolean> extends GenericPrompt<T> {
  toRay?: boolean;
}
