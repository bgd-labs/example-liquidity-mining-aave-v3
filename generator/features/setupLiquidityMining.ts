import * as addressBook from '@bgd-labs/aave-address-book';
import {Hex} from 'viem';
import {CodeArtifact, FEATURE, FeatureModule} from '../types';
import {LiquidityMiningSetup} from './types';
import {
  supplyUnderlyingAssetsSelectPrompt,
  supplyBorrowAssetsSelectPrompt,
  translateAssetToOracleLibUnderlying,
  translateAssetToAssetLibUnderlying,
  translateSupplyBorrowAssetToWhaleConstant,
} from '../prompts/assetsSelectPrompt';
import {addressPrompt} from '../prompts/addressPrompt';
import {percentPrompt} from '../prompts/percentPrompt';
import {numberPromptInDays} from '../prompts/numberPrompt';
import {
  getTokenDecimals,
  getExplorerTokenHoldersLink,
  getAddressOfSupplyBorrowAsset,
  getPoolChain,
  CHAIN_TO_CHAIN_ID,
  calculateExpectedWhaleRewards,
} from '../common';

export async function fetchLiquidityMiningSetupParams({pool}): Promise<LiquidityMiningSetup> {
  let rewardToken = await supplyUnderlyingAssetsSelectPrompt({
    message: 'Select the reward asset for the LM:',
    pool,
    required: true,
  });
  let rewardOracle: string;
  let rewardTokenAddress: Hex;

  if (rewardToken == 'custom') {
    rewardToken = await addressPrompt({
      message: 'Enter the address of the reward asset:',
      required: true,
    });
    rewardOracle = await addressPrompt({
      message: 'Enter the address of the reward oracle:',
      required: true,
    });
    rewardTokenAddress = rewardToken as Hex;
  } else {
    rewardOracle = translateAssetToOracleLibUnderlying(rewardToken, pool);
    rewardTokenAddress = rewardToken.includes('_aToken')
      ? addressBook[pool].ASSETS[rewardToken.replace('_aToken', '')].A_TOKEN
      : addressBook[pool].ASSETS[rewardToken].UNDERLYING;
    rewardToken = translateAssetToAssetLibUnderlying(rewardToken, pool);
  }
  const emissionsAdmin = await addressPrompt({
    message: 'Enter the address of the emissionsAdmin:',
    required: true,
  });

  const distributionEnd = await numberPromptInDays({
    message: 'Enter the total distribution time for the LM in days:',
    required: true,
  });
  const transferStrategy = await addressPrompt({
    message: 'Enter the address of the transfer strategy contract deployed:',
    required: true,
  });
  const assets = await supplyBorrowAssetsSelectPrompt({
    message: 'Enter the assets for which the LM should be configured:',
    pool,
    required: true,
  });

  const rewardAmounts: string[] = [];
  const whaleAddresses: Hex[] = [];
  const whaleExpectedRewards: string[] = [];
  let totalReward: number = 0;
  const chainId: number = CHAIN_TO_CHAIN_ID[getPoolChain(pool)];
  const rewardTokenDecimals = await getTokenDecimals(rewardTokenAddress, chainId);

  for (const index in assets) {
    const rewardAmount = await percentPrompt({
      message: `Enter the reward amount (in token units) for the reward token to be distributed for ${assets[index]}`,
      required: true,
    });
    totalReward += Number(rewardAmount);
    const supplyBorrowAssetAddress = getAddressOfSupplyBorrowAsset(pool, assets[index]);
    const whaleAddress = await addressPrompt({
      message: `Enter the whale address to test rewards for ${
        assets[index]
      } from ${getExplorerTokenHoldersLink(chainId, supplyBorrowAssetAddress)} `,
      required: true,
    });
    rewardAmounts.push(rewardAmount);
    whaleAddresses.push(whaleAddress);
    whaleExpectedRewards.push(
      await calculateExpectedWhaleRewards(
        whaleAddress,
        supplyBorrowAssetAddress,
        rewardAmount,
        chainId
      )
    );
  }

  return {
    emissionsAdmin,
    rewardToken,
    rewardTokenDecimals,
    rewardOracle,
    assets,
    distributionEnd,
    transferStrategy,
    rewardAmounts,
    totalReward,
    whaleAddresses,
    whaleExpectedRewards,
  };
}

export const setupLiquidityMining: FeatureModule<LiquidityMiningSetup> = {
  value: FEATURE.SETUP_LM,
  description: 'Setup new liquidity mining',
  async cli({pool}) {
    console.log(`Fetching information for setting new liquidity mining on ${pool}`);
    const response: LiquidityMiningSetup = await fetchLiquidityMiningSetupParams({pool});
    return response;
  },
  build({pool, cfg}) {
    const response: CodeArtifact = {
      code: {
        constants: [
          cfg.rewardToken.includes('0x')
            ? `address public constant override REWARD_ASSET = ${cfg.rewardToken};`
            : `address public constant override REWARD_ASSET = ${pool}Assets.${cfg.rewardToken}_UNDERLYING;`,
          `uint88 constant DURATION_DISTRIBUTION = ${cfg.distributionEnd} days;`,
          `uint256 public constant override TOTAL_DISTRIBUTION = ${cfg.totalReward} * 10 ** ${cfg.rewardTokenDecimals};`,
          `address constant EMISSION_ADMIN = ${cfg.emissionsAdmin};\n`,
          `address public constant override DEFAULT_INCENTIVES_CONTROLLER = ${pool}.DEFAULT_INCENTIVES_CONTROLLER;\n`,
          `ITransferStrategyBase public constant override TRANSFER_STRATEGY = ITransferStrategyBase(${cfg.transferStrategy});\n`,
          `IEACAggregatorProxy public constant override REWARD_ORACLE = IEACAggregatorProxy(${cfg.rewardOracle});\n`,
          ...cfg.assets.map((asset, index) => {
            let whaleConstants = `address constant ${translateSupplyBorrowAssetToWhaleConstant(
              asset,
              pool
            )} = ${cfg.whaleAddresses[index]};`;
            return whaleConstants;
          }),
        ],
        fn: [
          `
          function test_activation() public {
            vm.prank(EMISSION_ADMIN);
            IEmissionManager(${pool}.EMISSION_MANAGER).configureAssets(_getAssetConfigs());

            emit log_named_bytes(
              'calldata to submit from Gnosis Safe',
              abi.encodeWithSelector(
                IEmissionManager(${pool}.EMISSION_MANAGER).configureAssets.selector,
                _getAssetConfigs()
              )
            );

            ${cfg.assets
              .map(
                (assets, ix) => `
                _testClaimRewardsForWhale(
                  ${translateSupplyBorrowAssetToWhaleConstant(assets, pool)},
                  ${translateAssetToAssetLibUnderlying(assets, pool)},
                  DURATION_DISTRIBUTION,
                  ${cfg.whaleExpectedRewards[ix]} * 10 ** ${cfg.rewardTokenDecimals}
                );`
              )
              .join('\n')}
          }

            function _getAssetConfigs() internal override view returns (RewardsDataTypes.RewardsConfigInput[] memory) {
              uint32 distributionEnd = uint32(block.timestamp + DURATION_DISTRIBUTION);

              EmissionPerAsset[] memory emissionsPerAsset = _getEmissionsPerAsset();

              RewardsDataTypes.RewardsConfigInput[]
                memory configs = new RewardsDataTypes.RewardsConfigInput[](emissionsPerAsset.length);
              for (uint256 i = 0; i < emissionsPerAsset.length; i++) {
                configs[i] = RewardsDataTypes.RewardsConfigInput({
                  emissionPerSecond: _toUint88(emissionsPerAsset[i].emission / DURATION_DISTRIBUTION),
                  totalSupply: 0, // IMPORTANT this will not be taken into account by the contracts, so 0 is fine
                  distributionEnd: distributionEnd,
                  asset: emissionsPerAsset[i].asset,
                  reward: REWARD_ASSET,
                  transferStrategy: TRANSFER_STRATEGY,
                  rewardOracle: REWARD_ORACLE
                });
              }

              return configs;
            }

           function _getEmissionsPerAsset() internal override pure returns (EmissionPerAsset[] memory) {
            EmissionPerAsset[] memory emissionsPerAsset = new EmissionPerAsset[](${
              cfg.assets.length
            });
            ${cfg.assets
              .map(
                (assets, ix) => `
                emissionsPerAsset[${ix}] = EmissionPerAsset({
                  asset: ${translateAssetToAssetLibUnderlying(assets, pool)},
                  emission: ${cfg.rewardAmounts[ix]} * 10 ** ${cfg.rewardTokenDecimals}
                });`
              )
              .join('\n')}

            uint256 totalDistribution;
            for (uint256 i = 0; i < emissionsPerAsset.length; i++) {
              totalDistribution += emissionsPerAsset[i].emission;
            }
            require(totalDistribution == TOTAL_DISTRIBUTION, 'INVALID_SUM_OF_EMISSIONS');

            return emissionsPerAsset;
          }
          `,
        ],
      },
    };
    return response;
  },
};
