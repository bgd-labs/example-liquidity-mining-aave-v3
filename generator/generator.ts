import fs from 'fs';
import path from 'path';
import {generateContractName, generateFolderName} from './common';
import {liquidityMiningSetupTemplate} from './templates/liquiditymining.setup.template';
import {liquidityMiningUpdateTemplate} from './templates/liquiditymining.update.template';
import {confirm} from '@inquirer/prompts';
import {ConfigFile, Options, PoolConfigs, PoolIdentifier, Files} from './types';
import prettier from 'prettier';

const prettierSolCfg = await prettier.resolveConfig('foo.sol');
const prettierTsCfg = await prettier.resolveConfig('foo.ts');

/**
 * Generates all the file contents for aip/tests/payloads & script
 * @param options
 * @param poolConfigs
 * @returns
 */
export async function generateFiles(options: Options, poolConfigs: PoolConfigs): Promise<Files> {
  const jsonConfig = await prettier.format(
    `import {ConfigFile} from '../../generator/types';
    export const config: ConfigFile = ${JSON.stringify({
      rootOptions: options,
      poolOptions: (Object.keys(poolConfigs) as PoolIdentifier[]).reduce((acc, pool) => {
        acc[pool] = {configs: poolConfigs[pool]!.configs, cache: poolConfigs[pool]!.cache};
        return acc;
      }, {}),
    } as ConfigFile)}`,
    {...prettierTsCfg, filepath: 'foo.ts'}
  );

  async function createPayloadTest(options: Options, pool: PoolIdentifier) {
    const contractName = generateContractName(options, pool);
    const isLMSetup = options.feature == 'SETUP_LM';

    return {
      pool,
      payloadTest: await prettier.format(
        isLMSetup
          ? liquidityMiningSetupTemplate(options, poolConfigs[pool]!, pool)
          : liquidityMiningUpdateTemplate(options, poolConfigs[pool]!, pool),
        {
          ...prettierSolCfg,
          filepath: 'foo.sol',
        }
      ),
      contractName: contractName,
    };
  }

  return {
    jsonConfig,
    payloadTest: await createPayloadTest(options, options.pool),
  };
}

async function askBeforeWrite(options: Options, path: string, content: string) {
  if (!options.force && fs.existsSync(path)) {
    const currentContent = fs.readFileSync(path, {encoding: 'utf8'});
    // skip if content did not change
    if (currentContent === content) return;
    const force = await confirm({
      message: `A file already exists at ${path} do you want to overwrite`,
      default: false,
    });
    if (!force) return;
  }
  fs.writeFileSync(path, content);
}

/**
 * Writes the files according to defined folder/file format
 */
export async function writeFiles(options: Options, {jsonConfig, payloadTest}: Files) {
  const baseName = generateFolderName(options);
  const baseFolder = path.join(process.cwd(), 'tests', baseName);

  if (fs.existsSync(baseFolder)) {
    if (!options.force && fs.existsSync(baseFolder)) {
      const force = await confirm({
        message: 'A proposal already exists at that location, do you want to continue?',
        default: false,
      });
      if (!force) return;
    }
  } else {
    fs.mkdirSync(baseFolder, {recursive: true});
  }

  // write config
  await askBeforeWrite(options, path.join(baseFolder, 'config.ts'), jsonConfig);
  // write payloadTest
  await askBeforeWrite(
    options,
    path.join(baseFolder, `${payloadTest.contractName}.t.sol`),
    payloadTest.payloadTest
  );
}
