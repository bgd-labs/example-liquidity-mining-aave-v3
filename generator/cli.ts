import path from 'path';
import {Command, Option} from 'commander';
import {CHAIN_TO_CHAIN_ID, getDate, getPoolChain, pascalCase} from './common';
import {input, select} from '@inquirer/prompts';
import {ConfigFile, FEATURE, Options, POOLS, PoolCache, PoolConfigs, PoolIdentifier} from './types';
import {generateFiles, writeFiles} from './generator';
import {CHAIN_ID_CLIENT_MAP} from '@bgd-labs/js-utils';
import {getBlockNumber} from 'viem/actions';
import {setupLiquidityMining} from './features/setupLiquidityMining';
import {updateLiquidityMining} from './features/updateLiquidityMining';
import {LiquidityMiningSetup, LiquidityMiningUpdate} from './features/types';

const program = new Command();

program
  .name('aave-liquidity-mining-generator')
  .description('CLI to generate scripts for activating LM on aave')
  .version('1.0.0')
  .addOption(new Option('-f, --force', 'force creation (might overwrite existing files)'))
  .addOption(new Option('-c, --configFile <string>', 'path to config file'))
  .allowExcessArguments(false)
  .parse(process.argv);

let options = program.opts<Options>();
let poolConfigs: PoolConfigs = {};

const FEATURE_MODULES = [setupLiquidityMining, updateLiquidityMining];

async function generateDeterministicPoolCache(pool: PoolIdentifier): Promise<PoolCache> {
  const chain = getPoolChain(pool);
  const client = CHAIN_ID_CLIENT_MAP[CHAIN_TO_CHAIN_ID[chain]];
  return {blockNumber: Number(await getBlockNumber(client))};
}

async function fetchLMSetupOptions(pool: PoolIdentifier) {
  poolConfigs[pool] = {
    configs: {},
    artifacts: [],
    cache: await generateDeterministicPoolCache(pool),
  };

  const feature = FEATURE.SETUP_LM;
  const module = setupLiquidityMining;
  poolConfigs[pool]!.configs[feature] = await module.cli({
    options,
    pool,
    cache: poolConfigs[pool]!.cache,
  });
  poolConfigs[pool]!.artifacts.push(
    module.build({
      options,
      pool,
      cfg: poolConfigs[pool]!.configs[feature],
      cache: poolConfigs[pool]!.cache,
    })
  );
}

async function fetchLMUpdateOptions(pool: PoolIdentifier) {
  poolConfigs[pool] = {
    configs: {},
    artifacts: [],
    cache: await generateDeterministicPoolCache(pool),
  };

  const feature = FEATURE.UPDATE_LM;
  const module = updateLiquidityMining;
  poolConfigs[pool]!.configs[feature] = await module.cli({
    options,
    pool,
    cache: poolConfigs[pool]!.cache,
  });
  poolConfigs[pool]!.artifacts.push(
    module.build({
      options,
      pool,
      cfg: poolConfigs[pool]!.configs[feature],
      cache: poolConfigs[pool]!.cache,
    })
  );
}

if (options.configFile) {
  const {config: cfgFile}: {config: ConfigFile} = await import(
    path.join(process.cwd(), options.configFile)
  );
  options = {...options, ...cfgFile.rootOptions};
  poolConfigs = cfgFile.poolOptions as any;
  poolConfigs[options.pool]!.artifacts = [];

  if (options.feature == FEATURE.SETUP_LM) {
    poolConfigs[options.pool]!.artifacts.push(
      setupLiquidityMining.build({
        options,
        pool: options.pool,
        cfg: poolConfigs[options.pool]!.configs[FEATURE.SETUP_LM] as LiquidityMiningSetup,
        cache: poolConfigs[options.pool]!.cache,
      })
    );
  } else {
    poolConfigs[options.pool]!.artifacts.push(
      updateLiquidityMining.build({
        options,
        pool: options.pool,
        cfg: poolConfigs[options.pool]!.configs[FEATURE.UPDATE_LM] as LiquidityMiningUpdate,
        cache: poolConfigs[options.pool]!.cache,
      })
    );
  }
} else {
  options.feature = await select({
    message: 'Do you wish to setup a new liquidity mining or update an existing one?',
    choices: FEATURE_MODULES.map((m) => ({name: m.description, value: m.value})),
  });

  options.pool = await select({
    message: 'Select the Aave Pool:',
    choices: POOLS.map((v) => ({name: v, value: v})),
  });

  if (!options.title) {
    options.title = await input({
      message: 'Short title of the liquidity mining program',
      validate(input) {
        if (input.length == 0) return "Your title can't be empty";
        if (input.trim().length > 80) return 'Your title is to long';
        return true;
      },
    });
  }
  options.shortName = pascalCase(options.title);
  options.date = getDate();

  if (options.feature === FEATURE.SETUP_LM) {
    await fetchLMSetupOptions(options.pool);
  } else if (options.feature === FEATURE.UPDATE_LM) {
    await fetchLMUpdateOptions(options.pool);
  }
}

try {
  const files = await generateFiles(options, poolConfigs);
  await writeFiles(options, files);
} catch (e) {
  console.log(JSON.stringify({options, poolConfigs}, null, 2));
  throw e;
}
