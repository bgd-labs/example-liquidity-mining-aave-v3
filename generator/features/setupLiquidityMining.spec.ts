import {expect, describe, it} from 'vitest';
import {MOCK_OPTIONS_SETUP, liquidityMiningSetupConfig} from './mocks/configs';
import {generateFiles} from '../generator';
import {FEATURE, PoolConfigs} from '../types';
import {setupLiquidityMining} from './setupLiquidityMining';

describe('feature: setupLiquidityMining', () => {
  it('should return reasonable code', () => {
    const output = setupLiquidityMining.build({
      options: MOCK_OPTIONS_SETUP,
      pool: 'AaveV3EthereumLido',
      cfg: liquidityMiningSetupConfig,
      cache: {blockNumber: 42},
    });
    expect(output).toMatchSnapshot();
  });

  it('should properly generate files', async () => {
    const poolConfigs: PoolConfigs = {
      [MOCK_OPTIONS_SETUP.pool]: {
        pool: MOCK_OPTIONS_SETUP.pool,
        artifacts: [
          setupLiquidityMining.build({
            options: MOCK_OPTIONS_SETUP,
            pool: 'AaveV3EthereumLido',
            cfg: liquidityMiningSetupConfig,
            cache: {blockNumber: 42},
          }),
        ],
        configs: {[FEATURE.SETUP_LM]: liquidityMiningSetupConfig},
        cache: {blockNumber: 42},
      },
    };
    const files = await generateFiles(MOCK_OPTIONS_SETUP, poolConfigs);
    expect(files).toMatchSnapshot();
  });
});
