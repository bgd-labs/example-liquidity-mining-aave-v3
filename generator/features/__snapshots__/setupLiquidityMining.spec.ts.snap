// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feature: setupLiquidityMining > should properly generate files 1`] = `
{
  "jsonConfig": "import {ConfigFile} from '../../generator/types';
export const config: ConfigFile = {
  rootOptions: {
    pool: 'AaveV3EthereumLido',
    title: 'test',
    shortName: 'Test',
    date: '20231023',
    feature: 'SETUP_LM',
  },
  poolOptions: {
    AaveV3EthereumLido: {
      configs: {
        SETUP_LM: {
          emissionsAdmin: '0xac140648435d03f784879cd789130F22Ef588Fcd',
          rewardToken: 'AaveV3EthereumLidoAssets.wstETH_A_TOKEN',
          rewardTokenDecimals: 18,
          rewardOracle: 'AaveV3EthereumLidoAssets.wstETH_ORACLE',
          assets: ['wstETH_aToken'],
          distributionEnd: '14',
          transferStrategy: '0x0605a898535E9116Ff820347c536E3442F216Eb8',
          rewardAmounts: ['100'],
          totalReward: 100,
          whaleAddresses: ['0x07833EAdF87CD3079da281395f2fBA24b61F90f7'],
          whaleExpectedRewards: ['21.6'],
        },
      },
      cache: {blockNumber: 42},
    },
  },
};
",
  "payloadTest": {
    "contractName": "AaveV3EthereumLido_LMSetupTest_20231023",
    "payloadTest": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {AaveV3EthereumLido, AaveV3EthereumLidoAssets} from 'aave-address-book/AaveV3EthereumLido.sol';
import {IEmissionManager, ITransferStrategyBase, RewardsDataTypes, IEACAggregatorProxy} from '../../src/interfaces/IEmissionManager.sol';
import {LMSetupBaseTest} from '../utils/LMSetupBaseTest.sol';

contract AaveV3EthereumLido_LMSetupTest_20231023 is LMSetupBaseTest {
  address public constant override REWARD_ASSET =
    AaveV3EthereumLidoAssets.AaveV3EthereumLidoAssets.wstETH_A_TOKEN_UNDERLYING;
  uint88 constant DURATION_DISTRIBUTION = 14 days;
  uint256 public constant override TOTAL_DISTRIBUTION = 100 * 10 ** 18;
  address constant EMISSION_ADMIN = 0xac140648435d03f784879cd789130F22Ef588Fcd;

  address public constant override DEFAULT_INCENTIVES_CONTROLLER =
    AaveV3EthereumLido.DEFAULT_INCENTIVES_CONTROLLER;

  ITransferStrategyBase public constant override TRANSFER_STRATEGY =
    ITransferStrategyBase(0x0605a898535E9116Ff820347c536E3442F216Eb8);

  IEACAggregatorProxy public constant override REWARD_ORACLE =
    IEACAggregatorProxy(AaveV3EthereumLidoAssets.wstETH_ORACLE);

  address constant awstETH_WHALE = 0x07833EAdF87CD3079da281395f2fBA24b61F90f7;

  function setUp() public {
    vm.createSelectFork(vm.rpcUrl('mainnet'), 42);
  }

  function test_activation() public {
    vm.prank(EMISSION_ADMIN);
    IEmissionManager(AaveV3EthereumLido.EMISSION_MANAGER).configureAssets(_getAssetConfigs());

    emit log_named_bytes(
      'calldata to submit from Gnosis Safe',
      abi.encodeWithSelector(
        IEmissionManager(AaveV3EthereumLido.EMISSION_MANAGER).configureAssets.selector,
        _getAssetConfigs()
      )
    );

    _testClaimRewardsForWhale(
      awstETH_WHALE,
      AaveV3EthereumLidoAssets.wstETH_A_TOKEN,
      DURATION_DISTRIBUTION,
      21.6 * 10 ** 18
    );
  }

  function _getAssetConfigs()
    internal
    view
    override
    returns (RewardsDataTypes.RewardsConfigInput[] memory)
  {
    uint32 distributionEnd = uint32(block.timestamp + DURATION_DISTRIBUTION);

    EmissionPerAsset[] memory emissionsPerAsset = _getEmissionsPerAsset();

    RewardsDataTypes.RewardsConfigInput[]
      memory configs = new RewardsDataTypes.RewardsConfigInput[](emissionsPerAsset.length);
    for (uint256 i = 0; i < emissionsPerAsset.length; i++) {
      configs[i] = RewardsDataTypes.RewardsConfigInput({
        emissionPerSecond: _toUint88(emissionsPerAsset[i].emission / DURATION_DISTRIBUTION),
        totalSupply: 0, // IMPORTANT this will not be taken into account by the contracts, so 0 is fine
        distributionEnd: distributionEnd,
        asset: emissionsPerAsset[i].asset,
        reward: REWARD_ASSET,
        transferStrategy: TRANSFER_STRATEGY,
        rewardOracle: REWARD_ORACLE
      });
    }

    return configs;
  }

  function _getEmissionsPerAsset() internal pure override returns (EmissionPerAsset[] memory) {
    EmissionPerAsset[] memory emissionsPerAsset = new EmissionPerAsset[](1);

    emissionsPerAsset[0] = EmissionPerAsset({
      asset: AaveV3EthereumLidoAssets.wstETH_A_TOKEN,
      emission: 100 * 10 ** 18
    });

    uint256 totalDistribution;
    for (uint256 i = 0; i < emissionsPerAsset.length; i++) {
      totalDistribution += emissionsPerAsset[i].emission;
    }
    require(totalDistribution == TOTAL_DISTRIBUTION, 'INVALID_SUM_OF_EMISSIONS');

    return emissionsPerAsset;
  }
}
",
    "pool": "AaveV3EthereumLido",
  },
}
`;

exports[`feature: setupLiquidityMining > should return reasonable code 1`] = `
{
  "code": {
    "constants": [
      "address public constant override REWARD_ASSET = AaveV3EthereumLidoAssets.AaveV3EthereumLidoAssets.wstETH_A_TOKEN_UNDERLYING;",
      "uint88 constant DURATION_DISTRIBUTION = 14 days;",
      "uint256 public constant override TOTAL_DISTRIBUTION = 100 * 10 ** 18;",
      "address constant EMISSION_ADMIN = 0xac140648435d03f784879cd789130F22Ef588Fcd;
",
      "address public constant override DEFAULT_INCENTIVES_CONTROLLER = AaveV3EthereumLido.DEFAULT_INCENTIVES_CONTROLLER;
",
      "ITransferStrategyBase public constant override TRANSFER_STRATEGY = ITransferStrategyBase(0x0605a898535E9116Ff820347c536E3442F216Eb8);
",
      "IEACAggregatorProxy public constant override REWARD_ORACLE = IEACAggregatorProxy(AaveV3EthereumLidoAssets.wstETH_ORACLE);
",
      "address constant awstETH_WHALE = 0x07833EAdF87CD3079da281395f2fBA24b61F90f7;",
    ],
    "fn": [
      "
          function test_activation() public {
            vm.prank(EMISSION_ADMIN);
            IEmissionManager(AaveV3EthereumLido.EMISSION_MANAGER).configureAssets(_getAssetConfigs());

            emit log_named_bytes(
              'calldata to submit from Gnosis Safe',
              abi.encodeWithSelector(
                IEmissionManager(AaveV3EthereumLido.EMISSION_MANAGER).configureAssets.selector,
                _getAssetConfigs()
              )
            );

            
                _testClaimRewardsForWhale(
                  awstETH_WHALE,
                  AaveV3EthereumLidoAssets.wstETH_A_TOKEN,
                  DURATION_DISTRIBUTION,
                  21.6 * 10 ** 18
                );
          }

            function _getAssetConfigs() internal override view returns (RewardsDataTypes.RewardsConfigInput[] memory) {
              uint32 distributionEnd = uint32(block.timestamp + DURATION_DISTRIBUTION);

              EmissionPerAsset[] memory emissionsPerAsset = _getEmissionsPerAsset();

              RewardsDataTypes.RewardsConfigInput[]
                memory configs = new RewardsDataTypes.RewardsConfigInput[](emissionsPerAsset.length);
              for (uint256 i = 0; i < emissionsPerAsset.length; i++) {
                configs[i] = RewardsDataTypes.RewardsConfigInput({
                  emissionPerSecond: _toUint88(emissionsPerAsset[i].emission / DURATION_DISTRIBUTION),
                  totalSupply: 0, // IMPORTANT this will not be taken into account by the contracts, so 0 is fine
                  distributionEnd: distributionEnd,
                  asset: emissionsPerAsset[i].asset,
                  reward: REWARD_ASSET,
                  transferStrategy: TRANSFER_STRATEGY,
                  rewardOracle: REWARD_ORACLE
                });
              }

              return configs;
            }

           function _getEmissionsPerAsset() internal override pure returns (EmissionPerAsset[] memory) {
            EmissionPerAsset[] memory emissionsPerAsset = new EmissionPerAsset[](1);
            
                emissionsPerAsset[0] = EmissionPerAsset({
                  asset: AaveV3EthereumLidoAssets.wstETH_A_TOKEN,
                  emission: 100 * 10 ** 18
                });

            uint256 totalDistribution;
            for (uint256 i = 0; i < emissionsPerAsset.length; i++) {
              totalDistribution += emissionsPerAsset[i].emission;
            }
            require(totalDistribution == TOTAL_DISTRIBUTION, 'INVALID_SUM_OF_EMISSIONS');

            return emissionsPerAsset;
          }
          ",
    ],
  },
}
`;
